Implemented Battleships Model and Endpoint

Zusemat opened the pull request.

Zusemat committed:
Implemented Battleships Model and Endpoint

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/State.kt:
```diff
@@ -0,0 +1,7 @@
+package de.hwrberlin.sweii.tablegames.battleships
+// Mein Gott bin ich bescheuert
```
dieser Kommentar ist hier nicht sinnvoll

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+
+    private fun isShipSunk(shipType: ShipType): Boolean {
+        val size = shipType.size
+        return board.flatten().count { it == ShipStatus.HIT } == size
```
dass hier kann auch wahr sein, wenn mehere verschiedene Schiffe getroffen wurden

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+
+    fun winner(userIds: List<Long>): Long? {
+        for (userId in userIds) {
+            if (board.flatten().all { cell -> cell == ShipStatus.MISS || cell == ShipStatus.EMPTY }) {
```
das board kann auch eine zelle mit hit haben

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+        return if (lastTurn != userId) opponentBoard else Array(10) { Array(10) { ShipStatus.EMPTY } }
+    }
+
+    fun resetGame() {
```
Das wird nicht gebraucht, wenn man ein neues Spiel braucht, dann wird ein neues Objekt erstellt.
Diese Methode bitte löschen.

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    private fun isStraightLine(startX: Int, startY: Int, endX: Int, endY: Int): Boolean {
+        // Check if the ship is placed in a straight line (horizontal or vertical)
```
kommentar ist unnötig

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    private fun isValidPlacement(startX: Int, startY: Int, endX: Int, endY: Int): Boolean {
+        // Check if the placement is within the board boundaries
```
wenn hier ein kommentar nötig ist, dann bitte die Methode umbenennen. Z.B. isInBounds()

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+        return result
+    }
+
+    // Other game logic functions
```
kommentar ist unnötig

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+    enum class ShipType(val size: Int) { DESTROYER(2), SUBMARINE(3), CRUISER(3), BATTLESHIP(4), CARRIER(5) }
+
+    fun placeShip(x: Int, y: Int, size: Int, isHorizontal: Boolean, userId: Long): Boolean {
+        if (lastTurn != userId || size <= 0) {
```
Du stellst hier sicher, dass immer nur die selbe Person diese Methode ausführen kann... Das macht keinen sinn, wofür hast du dann überhaupt lastTurn?

termitehuegel replied:
Es wird nicht beschränkt, wie viele Schiffe von einen Typ plaziert wurden bzw. plaziert werden dürfen

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+            }
+        }
+
+        lastTurn = userId
```
lastTurn wird hier auf sich selbst gesetzt, siehe oben

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+        val endX = if (isHorizontal) x else x + size - 1
+        val endY = if (isHorizontal) y + size - 1 else y
+
+        if (!isValidPlacement(x, y, endX, endY) || !isStraightLine(x, y, endX, endY)) {
```
Die schiffe dürfen nicht kollidieren

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+            return false
+        }
+
+        // Place the ship on the board
```
dieser kommentar ist nicht wichtig

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    fun attack(x: Int, y: Int, userId: Long): Boolean {
+        if (lastTurn != userId) {
+            return false
+        }
```
Hier kommt immer nur eine Person vorbei...

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+            }
+        }
+
+        lastTurn = userId
```
Dieses Statement verändert die userId nicht

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+        val winner = winner(userIds)
+        return when {
+            winner != null -> State.DECIDED
+            board.all { row -> row.all { cell -> cell == ShipStatus.MISS } } -> State.DRAW
```
Es ist absolut unmöglich, dass beim Schiffe versenken ein unendschieden herauskommt...

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+
+data class BattleshipsGameState(
+    var lastTurn: Long = -1,
+    val board: Array<Array<ShipStatus>> = Array(10) { Array(10) { ShipStatus.EMPTY } },
```
Du führst hier nur das board von einem Spieler, das Spiel hat aber 2 Spieler

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+data class BattleshipsGameState(
+    var lastTurn: Long = -1,
+    val board: Array<Array<ShipStatus>> = Array(10) { Array(10) { ShipStatus.EMPTY } },
+    val opponentBoard: Array<Array<ShipStatus>> = Array(10) { Array(10) { ShipStatus.EMPTY } },
```
Dieses Feld wird nie mit Schiffen gefüllt? Wo werden die Schiffe vom zweiten Spieler geführt?

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+    var lastTurn: Long = -1,
+    val board: Array<Array<ShipStatus>> = Array(10) { Array(10) { ShipStatus.EMPTY } },
+    val opponentBoard: Array<Array<ShipStatus>> = Array(10) { Array(10) { ShipStatus.EMPTY } },
+    val shipsSunk: MutableList<ShipType> = mutableListOf()
```
Dieses Feld wird für nichts sinnvolles genutzt...

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+                board[x][y] = ShipStatus.HIT
+                opponentBoard[x][y] = ShipStatus.HIT
+
+                // Check if the ship is sunk
```
Dieser kommentar bietet keinen Mehrwert

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/State.kt:
```diff
@@ -0,0 +1,7 @@
+enum class State {
+    RUNNING,
+    DECIDED,
+    DRAW
```
Es gibt in diesem Spiel keinen Gleichstand.

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsActionRequest.kt:
```diff
@@ -0,0 +1,14 @@
+
+package de.hwrberlin.sweii.tablegames.battleships
+
+data class BattleshipsActionRequest(
```
Löse das bitte mit mehreren Methoden und damit auch mehreren Request-Objekten.

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsActionType.kt:
```diff
@@ -0,0 +1,8 @@
+// BattleshipsActionType.kt
+
+package de.hwrberlin.sweii.tablegames.battleships
```
Das Package ist falsch, deine IDE sollte dir das aber auch mitteilen

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsActionType.kt:
```diff
@@ -0,0 +1,8 @@
+// BattleshipsActionType.kt
```
Dieser kommentar ist unnötig, der Dateiname ist ersichtlich

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsActionRequest.kt:
```diff
@@ -0,0 +1,14 @@
+// BattleshipsActionRequest.kt
+
+package de.hwrberlin.sweii.tablegames.battleships
```
Falsches Package

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsActionRequest.kt:
```diff
@@ -0,0 +1,14 @@
+// BattleshipsActionRequest.kt
```
Bitte keine Kommentare, die nichts liefern

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsActionType.kt:
```diff
@@ -0,0 +1,8 @@
+
+package de.hwrberlin.sweii.tablegames.battleships
+
+enum class BattleshipsActionType {
+    PLACE_SHIP,
+    ATTACK
+}
```
Diese komplette enum sollte es nicht geben.

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsEndpoint.kt:
```diff
@@ -0,0 +1,93 @@
+// BattleshipsEndpoint.kt
```
Kommentar entfernen

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsStateResponse.kt:
```diff
@@ -0,0 +1,14 @@
+    val playerBoard: Array<Array<BattleshipsGameState.ShipStatus>>,
+    val opponentBoard: Array<Array<BattleshipsGameState.ShipStatus>>,
+    val turn: Long,
+    val state: State, //Bin mit hier nicht sicher
```
Kommentar entfernen

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsEndpoint.kt:
```diff
@@ -0,0 +1,93 @@
+        val currentUser = session.users.find { user -> user.id == gameState.lastTurn }
+            ?: throw InvalidActionException("Invalid state: Current user not found in session.")
+        val turn: Int =
+            if (userIds.indexOf(gameState.lastTurn) == -1) 0 else (userIds.indexOf(gameState.lastTurn) + 1) % userIds.size
```
Das ist immer false, weil der currentUser sonst nicht gefunden wurde

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsEndpoint.kt:
```diff
@@ -0,0 +1,93 @@
+
+    @CrossOrigin(originPatterns = ["*"])
+    @ResponseStatus(HttpStatus.NO_CONTENT)
+    @PostMapping("/action")
```
Bitte in zwei Methoden Splitten.

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsEndpoint.kt:
```diff
@@ -0,0 +1,93 @@
+                if (gameState.placeShip(
+                        actionRequest.x,
+                        actionRequest.y,
+                        actionRequest.shipSize!!,
```
Hier sollten shipSize und isHorizintal nicht nullable sein

termitehuegel commented on source/rest/src/main/kotlin/de/hwrberlin/sweii/tablegames/rest/battleships/BattleshipsStateResponse.kt:
```diff
@@ -0,0 +1,14 @@
+// BattleshipsStateResponse.kt
```
Kommentar entfernen

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+
+import de.hwrberlin.sweii.tablegames.general.GameState
+
+data class BattleshipsGameState(
```
Diese Klasse wird nirgendwo instanziert

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/Battleships.kt:
```diff
@@ -0,0 +1,168 @@
+package de.hwrberlin.sweii.tablegames.battleships
```
Move das bitte in eine eigenen Modul

termitehuegel commented on source/core/tic-tac-toe/src/main/kotlin/de/hwrberlin/sweii/tablegames/battleships/State.kt:
```diff
@@ -0,0 +1,7 @@
+package de.hwrberlin.sweii.tablegames.battleships
```
Das gehört in ein Batteships-Modul aber definitf nicht zu TicTacToe

termitehuegel reviewed (changes requested):
Ich würde dich bitten das noch einmal zu überarbeiten.
Ich habe dir einige stellen angemerkt, die keinen Sinn machen, unnötig sind oder zummindest umvormuliert werden sollten.

Außerdem würde ich dich bitten deine Logik in ein Batteships-Modul zu verschieben.

Die Logik deiner Klasse must du auch überdenken... Denn im Moment wird nur der Spielstand von einem Spieler gespeichert aber das Spiel wird in der Regel mit zwei Spielern gespielt.

Zusemat committed:
Corrected comments

Zusemat committed:
Merge remote-tracking branch 'origin/dev' into Branch_f4db8baf

Zusemat committed:
Corrected Battleships.kt, added PlayerState.kt, ShipStatus.kt, ShipType.kt and State.kt

termitehuegel committed:
Merge pull request #30 from yxyx-github/Branch_f4db8baf

Branch f4db8baf

Zusemat committed:
Corrected POM files

Zusemat committed:
Added test Class for BattleshipsTest.kt

Zusemat committed:
Corrected Test file location

Zusemat committed:
+ Added Tests for Battleships.kt
+ Fixed Minor Bugs
- Test for winner doesn't work yet because isShunkShip function in Battleships.kt doesn't work properly yet

Zusemat committed:
Modified winner test and isSunkShip for debugging. Test isn't passing yet

Zusemat committed:
Finalized Tests and Debugging

Zusemat committed:
Merge pull request #33 from yxyx-github/feature/test-battleships

Feature/test battleships

termitehuegel committed:
Merge remote-tracking branch 'origin/dev' into feature/add-battleships

termitehuegel committed:
fixed typo
